File: algorithms/002.Add_Two_Numbers/Two_Numbers.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def addTwoNumbers(self, l1, l2):
        dummy = ListNode(0)
        current, carry = dummy, 0
        
        while l1 is not None or l2 is not None:
            val = carry
            if l1 is not None:
                val = val + l1.val
                l1 = l1.next
            if l2 is not None:
                val = val + l2.val
                l2 = l2.next
            
            carry, val = int(val/10), val%10
            current.next = ListNode(val)
            current = current.next
            
        if carry == 1:
            current.next = ListNode(1)
        return dummy.next

File: algorithms/002.Add_Two_Numbers/Two_Numbers.go
package problem0002

// ListNode xx
type ListNode struct {
	Val  int
	Next *ListNode
}

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	dummy := &ListNode{}
	carry := 0
	current := dummy

	for l1 != nil || l2 != nil || carry > 0 {
		val := carry

		if l1 != nil {
			val = val + l1.Val
			l1 = l1.Next
		}

		if l2 != nil {
			val = val + l2.Val
			l2 = l2.Next
		}

		carry = val / 10

		current.Next = &ListNode{Val: val % 10}
		current = current.Next
	}

	return dummy.Next
}
